### ISSUE TITLE
Importing an old view history (option 8) results in issues

### ISSUE THREAD

**Issue by Tkkg1994** [ChuckPa/PlexDBRepair#82]
Hello and thank you very much for providing this tool. It's god damn useful!

I only now found this useful tool and tried to use it to recover old view historys of plex into my new setup (server broke down etc.). 

Now there's a good thing: the merge works, all the old view history appears BUT it looks like we have double entries in the db. For example, I watched "The Americans" on my old server. On the new server I did of course mark them as "played" so I can start watching the series at the same episode.

This results in the following:

![grafik](https://github.com/ChuckPa/PlexDBRepair/assets/10393524/84f134c6-a7f8-4a08-afd8-4655ae069ea9)

It looks like it counts all episodes twice. The first time I actually watched it (on the old server) and the second time I marked it as watched on the new server. This results in minus 13 :D

Now this is kind of a problem... it would be cool to remove these double entries and only use the most recent ones. I tried to repair db, also clean up in plex itself, redownload metadata/rescan etc. Didn't help

I can provide you with my .db file if you wish. But only in private of course

Thanks again and wish you a nice day!

**Comment by ChuckPa** [comment]
The table has a `unique constraint` which prevents duplicates.

I think what's happening is , with Plex's  Cloud-based watch history,  this feature is colliding with it.

It's not possible for me to selectively import individual records.    I am very much considering complete removal of this.

My recommendation to you is either use  Cloud-based history tracking or DB-based.   You can't use both

**Comment by Tkkg1994** [comment]
> The table has a `unique constraint` which prevents duplicates.
> 
> I think what's happening is , with Plex's Cloud-based watch history, this feature is colliding with it.
> 
> In this case, there are 13 references to the series but the series was matched using a different Agent so it shows up as negative (not found).
> 
> When importing, I don't know what the PMS settings are for the Agents.
> 
> My recommendation to you is either use Cloud-based history tracking or DB-based. You can't use both
> 
> I'm considering removing this feature.

Thank you for the answer. However it's not exactly what's happening since:

1. I wouldn't need to import my Plex History if the cloud based tracking was enabled. It isn't in my case, it's fully locally on my database
2. When I mark all episodes as "unwatched" the counter goes back to 0. It only turns negative when it's watched by myself on the old database and marked as watched on the 2nd database. And then merged of course.

i'm digging around in the db to see if I can find anything useful.

**Comment by ChuckPa** [comment]
> When I mark all episodes as "unwatched" the counter goes back to 0. It only turns negative when it's watched by myself on the old database and marked as watched on the 2nd database. And then merged of course.

If it turns negative when you watch it,  independent of importing anything,   stop right there.

Scenario:
1.   Mark unwatched
2.   Observe no negative counter
3.   Play normally / mark as watched manually
4.   Observe negative count.

ALL STOP.   That's a DB corruption deeper than what I can help with.
The only choice is to use a backup DB  (full swap) from elsewhere.

**Comment by Tkkg1994** [comment]
> > When I mark all episodes as "unwatched" the counter goes back to 0. It only turns negative when it's watched by myself on the old database and marked as watched on the 2nd database. And then merged of course.
> 
> If it turns negative when you watch it, independent of importing anything, stop right there.
> 
> Scenario:
> 
>     1. Mark unwatched
> 
>     2. Observe no negative counter
> 
>     3. Play normally / mark as watched manually
> 
> 
> > It only turns negative when it's watched by myself on the old database
> 
>     5. Observe negative count.
> 
> 
> If this is what's happening then ALL STOP. That's a DB corruption deeper than what I can help with. The only choice is to use a backup DB (full swap) from elsewhere.
> 
> My tool works on structure and whole tables. It doesn't have C++, agents, or metadata backing it like PMS does.

So I found a potential answer for this issue:

https://technicalramblings.com/blog/migrating-view-history-between-two-plex-servers-avoiding-negative-unwatched-count/#part-two-removing-the-duplicates

I will test this out tomorrow on my database (of course with a backup) and if it works fine, I'll let you know. Then you can include the command in your tool! Will for sure help people

**Comment by ChuckPa** [comment]
I have the same SQL from our devs.
```
"DELETE FROM metadata_item_settings
WHERE id in (SELECT MIN(id) 
FROM metadata_item_settings 
GROUP BY guid HAVING COUNT(guid) > 1);"
```

When this was tested,  it didn't work in all cases.   
Maybe it  was those particular databases  but also maybe not.
(I'm not the sharpest database guy)

**Comment by Tkkg1994** [comment]
> I have the same SQL from our devs.
> 
> ```
> "DELETE FROM metadata_item_settings
> WHERE id in (SELECT MIN(id) 
> FROM metadata_item_settings 
> GROUP BY guid HAVING COUNT(guid) > 1);"
> ```
> 
> When this was tested, it didn't work in all cases. Maybe it was those particular databases but also maybe not. (I'm not the sharpest database guy)
> 
> If you look in DBRepair.sh , line #1274-1284, you'll see it's there but commented out. You're welcome to experiment with it.
> 
> There might be a case where I ask to "Purge duplicates?" after importing because of Managed Users. I'm open to ideas.

Sadly I tried different things and the minus never disappeared. I'm now in contact with the writer of the article, but currently I have no news for you.. Sorry

**Comment by ChuckPa** [comment]
You may have to purge the watch history where the count < 0.

Doing this has always been danagerous.

**Comment by Tkkg1994** [comment]
@ChuckPa 

I have fixed the issue with the minus counts. You can include it in your script and hopefully make some people happy. See the code below

`DELETE FROM metadata_item_settings
WHERE rowid NOT IN (
    SELECT MIN(rowid)
    FROM metadata_item_settings
    GROUP BY guid, account_id
);
`

Right now everything looks fine but when I see any problems I'll report back. 
Credits would be lovely😊

**Comment by ChuckPa** [comment]
@Tkkg1994 

Is this

> DELETE FROM metadata_item_settings WHERE rowid NOT IN ( SELECT MIN(rowid) FROM metadata_item_settings GROUP BY guid, account_id ); 

a complete replacement for

> "DELETE FROM metadata_item_settings
WHERE id in (SELECT MIN(id)
FROM metadata_item_settings
GROUP BY guid HAVING COUNT(guid) < 0);"

?

**Comment by Tkkg1994** [comment]
> @Tkkg1994
> 
> Is this
> 
> > DELETE FROM metadata_item_settings WHERE rowid NOT IN ( SELECT MIN(rowid) FROM metadata_item_settings GROUP BY guid, account_id );
> 
> a complete replacement for
> 
> > "DELETE FROM metadata_item_settings
> > WHERE id in (SELECT MIN(id)
> > FROM metadata_item_settings
> > GROUP BY guid HAVING COUNT(guid) < 0);"
> 
> ?
> 
> where the NOT IN is the key change ?

This command from me is a complete replacement, yes.

Not only the NOT IN, it's a complete rework as I search for matching guid and account_id and group them, then delete all but one entry

**Comment by ChuckPa** [comment]
@Tkkg1994 

Included in  v1.0.10

Thank you.
